arbitrator/prover/src/lib.rs

```diff
@@ -18,19 +18,25 @@ pub mod wavm;
 use crate::machine::{argument_data_to_inbox, Machine};
 use arbutil::PreimageType;
 use eyre::Result;
+use lru::LruCache;
 use machine::{get_empty_preimage_resolver, GlobalState, MachineStatus, PreimageResolver};
 use static_assertions::const_assert_eq;
 use std::{
     ffi::CStr,
+    num::NonZeroUsize,
     os::raw::{c_char, c_int},
     path::Path,
     sync::{
         atomic::{self, AtomicU8},
-        Arc,
+        Arc, RwLock,
     },
 };
 use utils::{Bytes32, CBytes};
 
+lazy_static::lazy_static! {
+    static ref CACHE_PREIMAGE_REHASH_CHECK: RwLock<LruCache<(Bytes32, PreimageType), bool>> = RwLock::new(LruCache::new(NonZeroUsize::new(1024).unwrap()));
```

Comment:
I think we should store the preimage in the LRU cache, so we can confirm that the right preimage is supplied in future invocations. I think we should also make this cache only for eth versioned hashes, and limit its size to 12 entries (enough to cover 2 batches each containing 6 blobs, which would be a total of ~1MB). Other preimage types aren't worth caching because they're fast to hash and not as repeated.

Reply:
I'd also make this a mutex instead of a RwLock. That way you can call get on the LRU cache which should also update the position of it in the LRU and remove the need to re-add it.

Reply:
Essentially we would be removing the need to call resolver to get the preimage, i think thats a good idea. 
Initially I wanted LRU cache to remove the need to recheck a preimage but this is seems like an added bonus